x-pack/plugin/analytics/src/main/java/org/elasticsearch/xpack/analytics/boxplot/InternalBoxplot.java

```diff
@@ -0,0 +1,200 @@
+ */
+
+package org.elasticsearch.xpack.analytics.boxplot;
+
+import org.elasticsearch.common.io.stream.StreamInput;
+import org.elasticsearch.common.io.stream.StreamOutput;
+import org.elasticsearch.common.xcontent.XContentBuilder;
+import org.elasticsearch.search.DocValueFormat;
+import org.elasticsearch.search.aggregations.InternalAggregation;
+import org.elasticsearch.search.aggregations.metrics.InternalNumericMetricsAggregation;
+import org.elasticsearch.search.aggregations.metrics.TDigestState;
+import org.elasticsearch.search.aggregations.pipeline.PipelineAggregator;
+
+import java.io.IOException;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+
+public class InternalBoxplot extends InternalNumericMetricsAggregation.MultiValue implements Boxplot {
+
+    enum Metrics {
+
+        min, max, q1, q2, q3;
+
+        public static Metrics resolve(String name) {
+            return Metrics.valueOf(name);
+        }
+    }
+
+    private final TDigestState state;
+
+    InternalBoxplot(String name, TDigestState state, DocValueFormat formatter, List<PipelineAggregator> pipelineAggregators,
+                    Map<String, Object> metaData) {
+        super(name, pipelineAggregators, metaData);
+        this.state = state;
+        this.format = formatter;
+    }
+
+    /**
+     * Read from a stream.
+     */
+    public InternalBoxplot(StreamInput in) throws IOException {
+        super(in);
+        format = in.readNamedWriteable(DocValueFormat.class);
+        state = TDigestState.read(in);
+    }
+
+    @Override
+    protected void doWriteTo(StreamOutput out) throws IOException {
+        out.writeNamedWriteable(format);
+        TDigestState.write(state, out);
+    }
+
+    @Override
+    public String getWriteableName() {
+        return BoxplotAggregationBuilder.NAME;
+    }
+
+    @Override
+    public double getMin() {
+        return state.getMin();
+    }
+
+    @Override
+    public double getMax() {
+        return state.getMax();
+    }
+
+    @Override
+    public double getQ1() {
+        return state.quantile(0.25);
+    }
+
+    @Override
+    public double getQ2() {
+        return state.quantile(0.5);
+    }
+
+    @Override
+    public double getQ3() {
+        return state.quantile(0.75);
+    }
+
+    @Override
+    public String getMinAsString() {
+        return valueAsString(Metrics.min.name());
+    }
+
+    @Override
+    public String getMaxAsString() {
+        return valueAsString(Metrics.max.name());
+    }
+
+    @Override
+    public String getQ1AsString() {
+        return valueAsString(Metrics.q1.name());
+    }
+
+    @Override
+    public String getQ2AsString() {
+        return valueAsString(Metrics.q2.name());
+    }
+
+    @Override
+    public String getQ3AsString() {
+        return valueAsString(Metrics.q3.name());
+    }
+
+    @Override
+    public double value(String name) {
+        Metrics metrics = Metrics.valueOf(name);
+        switch (metrics) {
```

Comment:
Would you be making a private method on `Metrics` that takes `this` as a parameter and calls the right method instead?